src/measure_theory/hausdorff_measure.lean

```diff
@@ -0,0 +1,474 @@
+  suffices : ∀ a, (∑' (k : ℕ), μ (S (2 * k + 1 + a) \ S (2 * k + a))) ≠ ∞,
+    from ⟨by simpa using this 0, by simpa using this 1⟩,
+  refine λ r, ne_top_of_le_ne_top htop _,
+  rw [← Union_S, ennreal.tsum_eq_supr_nat, supr_le_iff],
+  intro n,
+  rw [← hm.finset_Union_of_pairwise_separated],
+  { exact μ.mono (Union_subset $ λ i, Union_subset $ λ hi x hx, mem_Union.2 ⟨_, hx.1⟩) },
+  suffices : ∀ i  j, i < j → is_metric_separated (S (2 * i + 1 + r)) (s \ S (2 * j + r)),
+    from λ i _ j _ hij, hij.lt_or_lt.elim
+      (λ h, (this i j h).mono (inter_subset_left _ _) (λ x hx, ⟨hx.1.1, hx.2⟩))
+      (λ h, (this j i h).symm.mono  (λ x hx, ⟨hx.1.1, hx.2⟩) (inter_subset_left _ _)),
+  intros i j hj,
+  have A : ((↑(2 * j + r))⁻¹ : ℝ≥0∞) < (↑(2 * i + 1 + r))⁻¹,
+    by { rw [ennreal.inv_lt_inv, ennreal.coe_nat_lt_coe_nat], linarith },
+  refine ⟨(↑(2 * i + 1 + r))⁻¹ - (↑(2 * j + r))⁻¹, by simpa using A, λ x hx y hy, _⟩,
+  have : inf_edist y t < (↑(2 * j + r))⁻¹, from not_le.1 (λ hle, hy.2 ⟨hy.1, hle⟩),
+  rcases exists_edist_lt_of_inf_edist_lt this with ⟨z, hzt, hyz⟩,
+  have hxz : (↑(2 * i + 1 + r))⁻¹ ≤ edist x z, from le_inf_edist.1 hx.2 _ hzt,
+  apply ennreal.le_of_add_le_add_right (hyz.trans_le le_top),
+  refine le_trans _ (edist_triangle _ _ _),
+  refine (add_le_add le_rfl hyz.le).trans (eq.trans_le _ hxz),
+  rw [ennreal.sub_add_cancel_of_le A.le]
+end
+
+lemma le_caratheodory [measurable_space X] [borel_space X] (hm : is_metric μ) :
+  ‹_› ≤ μ.caratheodory :=
+by { rw @borel_space.measurable_eq X _ _, exact hm.borel_le_caratheodory }
+
+end is_metric
+
+/-!
+### Constructors of metric outer measures
+
+In this section we provide constructors `measure_theory.outer_measure.mk_metric'` and
+`measure_theory.outer_measure.mk_metric` and prove that these outer measures are metric outer
+measures. We also prove basic lemmas about `map`/`comap` of these measures.
+-/
+
+/-- Auxiliary definition for `outer_measure.mk_metric'`: given a function on sets
+`m : set X → ℝ≥0∞`, returns the maximal outer measure `μ` such that `μ s ≤ m s`
+for any set `s` of diamenter at most `r`.-/
+def mk_metric'.pre (m : set X → ℝ≥0∞) (r : ℝ≥0∞) :
+  outer_measure X :=
+bounded_by $ extend (λ s (hs : diam s ≤ r), m s)
+
+/-- Given a function `m : set X → ℝ≥0∞`, `mk_metric' m` is the supremum of `mk_metric'.pre m r`
+over `r > 0`. Equivalently, it is the limit of `mk_metric'.pre m r` as `r` tends to zero from
+the right. -/
+def mk_metric' (m : set X → ℝ≥0∞) :
+  outer_measure X :=
+⨆ r > 0, mk_metric'.pre m r
+
+/-- Given a function `m : ℝ≥0∞ → ℝ≥0∞`, `mk_metric m = mk_metric' (λ s, m (emetric.diam s))`
+is the supremum of `mk_metric'.pre (λ s, m (emetric.diam s)) r` over `r > 0`. Each
+`mk_metric'.pre (λ s, m (emetric.diam s)) r` is the maximal outer measure `μ` such that
+`μ s ≤ m (emetric.diam s)` for any set of diameter less than `r`. -/
+def mk_metric (m : ℝ≥0∞ → ℝ≥0∞) : outer_measure X :=
+mk_metric' (λ s, ⨆ (hs : ¬s.subsingleton), m (diam s))
```

Comment:
I understand that you exclude subsingletons here to avoid talking of the behavior of the function `r -> r ^ d` at `0`.  You should mention this choice in the docstring.